
```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '09-wright_fisher')
```

# Simulating evolution

In this lab, we'll build a simulation to explore genetic drift using the Wright-Fisher model.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Describe the phenomenon of genetic drift.
2. Explain why random draws from a binomial distribution are a good way to mimic the effect of drift.
3. Interpret allele frequency patterns that occur as a result of drift.
4. Write a for loop in R.
5. Write a function to run code multiple times with different parameters.


## Genetic drift

In all populations, **genetic drift** acts to change allele frequencies over time. Drift refers to random changes in an allele's frequency between generations. These random changes occur because individuals carrying different alleles will have different numbers of offspring due to chance.

<center>

![**Fig. 1 ([source](https://sites.google.com/a/winnebagoschools.org/mrsweber-science/articles-for-teaching/genetic-drift)).** An allele's frequency "drifts" between generations due to random chance.](09-wright_fisher/images/drift.png)

</center>

Drift differs from **selection**, which is a deterministic (non-random) change in an allele's frequency. If an allele is under selection, it's more likely to increase or decrease in frequency depending on whether it is beneficial or deleterious. Genetic drift, on the other hand, cannot consistently cause an allele's frequency to increase or decrease.


## The Wright-Fisher model

The **Wright-Fisher model** is one of the most commonly used models of genetic drift.

In this model, we assume that a population:

* Mates randomly
* Number of individuals remains constant between generations
* Today, we'll also assume that the population is _haploid_

Populations in the real world don't behave like an ideal Wright-Fisher population, so their **effective population size** `Ne` is usually much smaller than their actual population size. The effective population size of the human population is only [12,800--14,400](https://www.pnas.org/content/109/44/17758) individuals, even though its actual size is around 8 billion.


## Allele frequency, fixation, and loss

The Wright-Fisher model describes the behavior of a single **allele**, which can be at any variable site in a population (a SNP, insertion/deletion, version of a gene, etc.).

The allele of interest begins the simulation at some initial **allele frequency (AF)**. This allele frequency is the proportion of individuals in the population who carry that allele, and is always between 0 and 1.

An allele becomes **fixed** in a population if it reaches an allele frequency of 1, and is **lost** if it reaches a frequency of 0. At either of these points, it is no longer considered a variable site because either everyone or no one in the population carries it.

<center>

![**Fig. 2 ([source](https://sites.radford.edu/~rsheehy/Gen_flash/popgen/)).** Trajectories of alleles at two loci fixing, at `AF = 0` and `AF = 1`.](09-wright_fisher/images/fixation.png)

</center>


## Modeling allele frequencies

In the Wright-Fisher model, we track a population over the course of many generations. Within each generation:

* For every individual, we perform a coin flip to determine whether or not they have the allele.
* Unlike a coin, the probability of receiving an allele is equal to its **allele frequency** in the current generation.
  * The more common an allele is in this generation, the more likely it is that someone in the next generation will also carry it.

After flipping these coins, we know the number of people in the next generation who carry the allele.

<center>

![**Fig. 3.** Every individual flips a weighted coin to determine whether they will carry the blue allele in the next generation. The probability of carrying the allele is equal to the allele's frequency in the current generation.](09-wright_fisher/images/coin_flips.jpg){width=65%}

</center>


## The binomial distribution

Instead of having to actually simulate all the coin flips, we can get the number of allele carriers by performing a single draw from a **binomial distribution** with size `N` (# of individual) and success probability `p = AF`. This distribution tells you how many successes you expect to see from a set of `N` independent coin flips.

If we try to draw 100,000 times from a binomial distribution with population size `100` and success probability (AF) `0.5`, it will look something like this:

```{r, echo = FALSE, warning = FALSE, fig.align = 'center', fig.cap = "100,000 draws from a binomial distribution."}
set.seed(5)
library(ggplot2)
data = data.frame(num = rbinom(n = 100000, size = 1000, prob = 0.5))
ggplot(data, aes(x = num/10)) +
  geom_histogram(bins = 50, color = "#757575", fill = "#8dddeb") +
  labs(x = "Number of individuals with A allele",
       y = "How often we drew this number") +
  theme_bw() +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())
```

The majority of the distribution lies between 48 and 52. Just as we expect based on the allele frequency, the next generation will most likely have around 48-52 individuals with the `A` allele. But because this is a random draw, there's a small chance that we might end up with many more or many fewer than that number.


## Setup

### R packages

```{r, results = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
```

### Data

We'll simulate all of our own data for this lab!


## The `rbinom` function

The basis of our simulation is R's `rbinom` function, which allows us to sample from a binomial distribution. `rbinom` takes three arguments:

* `n`: how many times we're drawing from the distribution
* `size`: the size of the population we're sampling from (i.e. `N`)
* `p`: the success probability (i.e. allele frequency)

Every generation, we'll draw **once** to produce the number of individuals carrying the `A` allele in the next generation.

Let's once again look at a population of size 100, and an `A` allele currently at `AF = 0.5`. We use `rbinom` to get the number of individuals in the next generation who will have `A`:

```{r}
rbinom(n = 1, size = 100, prob = 0.5)
```

***
<details> <summary> Change the `rbinom` code so that it returns the allele frequency (instead of the number of individuals). </summary>

```{r}
# divide by the population size to get AF
rbinom(n = 1, size = 100, prob = 0.5) / 100
```

</details>
***

***
<details> <summary> Why do we get a different number every time we run `rbinom`? </summary>

`rbinom` generates a random number between 0 and 100. Because it's random, the number it draws will be different every time we run it.

</details>
***


## Increasing population size

Currently, we're drawing from a population of 100 individuals. Now let's see what happens when we increase the population size. (Feel free to run this code block multiple times!)

```{r}
rbinom(n = 1, size = 10000, prob = 0.5) / 10000
```

If you run the code block above multiple times, you'll observe that the AF is much closer to 0.5 than it was with a population of size 100. This lends to our intuition that an allele's frequency fluctuates much more when a population is small, and is more stable when the population size is large.

***
<details> <summary> How does population size affect an allele's time to fixation? </summary>

As population size gets larger, the allele will take longer to fix.

</details>
***


## Simulating multiple generations

Currently, we draw _once_ from a binomial distribution to get the number of individuals in one generation who carry the allele of interest. How do we adapt this to simulate multiple generations?

***
<details> <summary> Increasing `n` (ex: `rbinom(n = 10, size = 100, prob = 0.5)`) </summary>

Increasing `n` only gives you multiple replicate draws from the same distribution.

This doesn't reflect multiple generations, because the allele frequency doesn't update between generations based on the new number of alleles -- it uses `prob = 0.5` every time.

</details>
***


## For loops

Instead of drawing multiple times from the same distribution, we can write a **for loop** to repeatedly generate and update the number of individuals with the `A` allele.

A **for loop** allows you to run some code X number of times. For example:

```{r}
for (i in 1:3) {
  print(i)
}
```

This for loop goes through all the values between 1 and 3, and prints each of them out.

***
<details> <summary> Modify the for loop to instead run our `rbinom` command. </summary>

```{r}
for (i in 1:3) {
  print(rbinom(n = 1, size = 100, prob = 0.5) / 100)
}
```

</details>
***


## Updating variables within a for loop

We also need to update the allele frequency in every iteration of the for loop. We do this by adding a `freq` variable that keeps track of the current AF:

```{r}
# start an initial AF of 0.5
freq <- 0.5

for (i in 1:3) {
  # run rbinom to generate the AF for the next generation
  new_freq <- rbinom(n = 1, size = 100, prob = freq) / 100
  # print new AF
  print(new_freq)
  # update `freq` in each iteration of the loop
  freq <- new_freq
}
```


## Adding a population size variable

Using the `freq` variable also gives us more flexibility than hard-coding an allele frequency into the `rbinom` command.

***
<details> <summary> Add to your code so that we also provide `Ne` (effective population size) as a variable (_without_ updating it in the for loop). </summary>

```{r}
# set effective population size outside of for loop
Ne <- 100
# start an initial AF of 0.5
freq <- 0.5

for (i in 1:3) {
  # run rbinom to generate the AF for the next generation
  new_freq <- rbinom(n = 1, size = Ne, prob = freq) / Ne
  # print new AF
  print(new_freq)
  # update `freq` in each iteration of the loop
  freq <- new_freq
}
```

</details>
***


## Changes in AF over generations

Try increasing the number of generations we run the simulation for. What patterns of change do you observe in the allele frequencies?

```{r}
Ne <- 100
freq <- 0.5

for (i in 1:20) {
  new_freq <- rbinom(n = 1, size = Ne, prob = freq) / Ne
  print(new_freq)
  freq <- new_freq
}
```

The allele frequency starts approaching either 0 or 1.


## Storing AFs in a vector

To plot how the AF changes over time, we can store the AF at each generation in a **vector**.

Vectors R's version of a list, and are formed with the `c()` function, which stands for "combine":

```{r}
my_vec <- c(0.5, 0.6)
my_vec
```

You can append elements to a vector called `my_vec` by running: `my_vec <- c(my_vec, new_element)`.

***
<details> <summary> Modify the code block with our for loop to create a vector for storing allele frequencies, and then append the updated AF to it every generation. </summary>

We need to create the vector **before** the for loop, and then append to the vector **within** the for loop.

```{r}
Ne <- 100
freq <- 0.5
# create vector to store AFs in
freq_vector <- freq

for (i in 1:20) {
  new_freq <- rbinom(n = 1, size = Ne, prob = freq) / Ne
  # add new freq to the AF vector
  freq_vector <- c(freq_vector, new_freq)
  freq <- new_freq
}

freq_vector
```

</details>
***


## Reformatting AFs for plotting

Because `ggplot` requires its input data to be formatted as a table, we have to convert `freq_vector` into some form of table (ex: a tibble or dataframe).

```{r}
sim_results <- tibble(af = freq_vector)
head(sim_results)
```

This table contains the information that we want on the plot's y axis. We can now add in a column containing the plot's x axis data, which is the **generation** that each AF value corresponds to.

```{r}
sim_results <- tibble(af = freq_vector,
                      gen = 1:21)
head(sim_results)
```

***
<details> <summary> Why does the `gens` column range from 1 to 21 (instead of 1 to 20)? </summary>

We add our starting allele frequency to `freq_vector`, and then simulate for 20 generations. This means that we end up with 21 AFs in our vector.

</details>
***


## Plotting AF trajectory

***
<details> <summary> Plot the trajectory of AFs over time with `ggplot`. </summary>

```{r}
ggplot(data = sim_results,
       aes(x = gen, y = af)) +
  geom_line()
```

</details>
***


## Simulating different parameters with a function

It would be nice to be able to run our Wright-Fisher simulation with different parameters -- like different starting allele frequencies, population sizes, etc. -- without having to edit the for loop code every time. We can use a **function** to generalize the code above so we can easily re-run it.

***
<details> <summary> The structure of an R function </summary>

You've already encountered many functions in R, even if you didn't realize it at the time - `rbinom`, `ggplot`, and `print` are all examples of functions. An R function has [four parts](https://www.tutorialspoint.com/r/r_functions.htm):

```
<Name> <- function(<Argument(s)>) {
  <Body>
  <return()>
}
```

* **Name** − The function is stored in your R environment as an object with this name, and you use the name to call it
* **Argument(s)** − Optional; input values that the function performs operations on
* **Body** − The code that describes what the function does
* **Return** − Optional; a `return` statement allows the function to return a value to the user. Without a return statement, you won't be able to access the function's output

</details>
***

Here's an example function that takes in three parameters for running `rbinom`, and returns the output of `rbinom`.
```
binom_sim <- function(myN, mySize, myProb) {
  output <- rbinom(myN, mySize, myProb)
  return(output)
}
```

***
<details> <summary> How do I know when to use a function? </summary>

Functions are useful whenever you have code that you want to run multiple times with slightly different parameters. If you find yourself copying over code several times and changing just a few things, you should consider writing a function instead.

</details>
***


## Creating a Wright-Fisher function

We want our function to take in parameters for the starting allele frequency, population size, and number of generations to simulate. It should return the `sim_results` dataframe so that we can plot the allele frequency trajectory.

To write a function, we can put the code that we just wrote into the function body:

```{r}
run_sim <- function(Ne, freq, generations) {
  
  # note how we don't define our initial parameters for Ne, freq, etc.
  # because we're passing in those parameters as arguments
  
  freq_vector <- freq
  for (i in 1:generations) {
    new_freq <- rbinom(n = 1, size = Ne, prob = freq) / Ne
    freq_vector <- c(freq_vector, new_freq)
    freq <- new_freq
  }
  
  # convert vector of AFs into a tibble for plotting
  sim_results <- tibble(afs = freq_vector,
                        gen = 1:(generations+1))
  
  # return the tibble of AFs, so that we can access the results
  return(sim_results)
}
```


## Running a function

The code block we just wrote **defines** the function (i.e., tells R what it should do). Now we can **run** the function with parameters of our choosing and plot the output:

```{r}
# run function
results <- run_sim(Ne = 1000, freq = 0.5, generations = 10000)

# plot output
ggplot(data = results,
       aes(x = gen, y = afs)) +
  geom_line()
```

***
<details> <summary> Run your `run_sim` function a few times with different input population sizes and AFs. How does changing these inputs affect the AF trajectories that you see? </summary>

```{r}
# simulate a very large population
results <- run_sim(Ne = 100000, freq = 0.5, generations = 1000)

# plot output
ggplot(data = results,
       aes(x = gen, y = afs)) +
  geom_line()

# simulate a population with AF closer to 0
results <- run_sim(Ne = 1000, freq = 0.1, generations = 1000)

# plot output
ggplot(data = results,
       aes(x = gen, y = afs)) +
  geom_line()
```

In general, decreasing the population size or shifting the starting AF away from 0.5 will decrease the time to fixation for an allele.

</details>
***


## Conclusion

In this lab, we built a Wright-Fisher simulation for one allele, allowing us to track how we expect its frequency to change over time under the principles of genetic drift.

* Within each simulated generation, we drew the number of allele carriers in the next generation from a **binomial distribution**.
<br></br>
* We performed these AF draws within a **for loop**, allowing us to update the current AF with each new generation.
<br></br>
* Finally, we wrapped our Wright-Fisher simulation code into a **function**, allowing us to easily re-run our simulation with different parameters for the starting AF, population size, and number of generations to simulate for.
<br></br>
* We observed that smaller population sizes and more extreme AFs (closer to 0 or 1) generally decrease an allele's time to fixation.

This simple simulation forms the core of most models used in evolutionary genetics research, which often extend this model to simulate more complex phenomena (such as different forms of selection).


## Homework

One way to extend our simple Wright-Fisher model is to add in selection as a parameter. Selection affects our model by altering the probability of sampling our allele of interest each generation (e.g., positive selection increases the probability, and negative selection decreases it).

Previously, we assumed that this probability was equivalent to the allele's frequency, or $p = \frac{i}{N_e}$, where $N_e$ is the population size and $i$ is the number of individuals who carry the allele.

For the purposes of this homework, we assume that in a model with selection, this probability is instead:

$$
p = \frac{i(1 + s)}{N_e - i + i(1+s)}
$$

where $s$ is the **selection coefficient**, and ranges from `-1` to `1`.

***
<details> <summary> What does this probability become in the absence of selection (i.e., when $s = 0$)? </summary>

The probability becomes $\frac{i}{N_e}$, which is the same as the allele frequency.

</details>
***

#### Learning Objectives

* Practice writing functions in R
* Interpret allele frequency trajectories under selection and drift

#### Assignment

In the code block below, modify your `run_sim` function so that it takes in a selection coefficient `s` as a parameter. Run the simulation a few times with and without (`s = 0`) selection, but keeping other parameters the same (`Ne = 10000`, `freq = 0.5`, `generations = 10000`). What do you notice about the allele frequency trajectories?

Note that most selection coefficients are thought to be [extremely small](https://journals.plos.org/plosgenetics/article/figure?id=10.1371/journal.pgen.1004697.g002) -- the largest known selection coefficients in humans are around [0.05](https://elifesciences.org/articles/63177).

***
<details> <summary> Solution </summary>

```{r}
# simulation function with selection
run_sim_selection <- function(Ne, freq, generations, s) {
  
  freq_vector <- freq
  for (i in 1:generations) {
    # calculate p, the probability of sampling the allele, based on s
    i <- freq * Ne # number of individuals who currently carry the allele
    p <- i*(1+s) / (Ne - i + i*(1+s))
    
    # prob is now `p`, rather than `freq`
    new_freq <- rbinom(n = 1, size = Ne, prob = p) / Ne
    freq_vector <- c(freq_vector, new_freq)
    freq <- new_freq
  }
  
  # convert vector of AFs into a tibble for plotting
  sim_results <- tibble(afs = freq_vector,
                        gen = 1:(generations+1))
  
  # return the tibble of AFs, so that we can access the results
  return(sim_results)
}
```

Run and plot the simulation with selection:

```{r}
results <- run_sim_selection(Ne = 10000,
                             freq = 0.5,
                             generations = 10000,
                             s = -0.001)
ggplot() +
  geom_line(data = results, aes(x = gen, y = afs)) +
  ylim(0, 1) +
  ylab("Allele frequency") +
  xlab("Generation") +
  ggtitle("Simulation with selection") +
  theme(plot.title = element_text(hjust = 0.5)) # to center the title
```

Run and plot the simulation without selection:

```{r}
results <- run_sim_selection(Ne = 10000,
                             freq = 0.5,
                             generations = 10000,
                             s = 0)
ggplot() +
  geom_line(data = results, aes(x = gen, y = afs)) +
  ylim(0, 1) +
  ylab("Allele frequency") +
  xlab("Generation") +
  ggtitle("Simulation without selection") +
  theme(plot.title = element_text(hjust = 0.5)) # to center the title
```

We observe that selection tends to decrease the time it takes for an allele to either fix or go extinct, because it directionally biases the probability of sampling that allele. Decreasing the absolute value of the selection coefficient will make the simulation behave more like drift.

</details>
***